File: examples/python/advanced.py
#!/usr/bin/env python3
# Copyright (c) 2026 vectorless developers
# SPDX-License-Identifier: Apache-2.0

"""
Advanced usage example - Full Configuration.

This example demonstrates how to use a full configuration file
for advanced use cases where you need fine-grained control.

Usage:
    # First, copy the example config
    cp config.toml ./my_vectorless.toml

    # Edit my_vectorless.toml to customize settings

    # Run the example
    python advanced.py
"""

import os
import tempfile
from vectorless import Engine, IndexContext


def main():
    print("=== Vectorless Advanced Example (Full Configuration) ===\n")

    # Method 1: Use explicit config file path
    # This loads all settings from the specified config file
    engine = Engine(config_path="./config.toml")

    print("✓ Engine created with config file\n")

    # Index a document
    content = """
# Product Documentation

## Overview

This is a comprehensive guide for our product.

## Configuration

The system supports multiple configuration methods:

### 1. Zero Configuration
Just set OPENAI_API_KEY environment variable.

### 2. Environment Variables
- VECTORLESS_MODEL: Set default model
- VECTORLESS_ENDPOINT: Set API endpoint
- VECTORLESS_WORKSPACE: Set workspace directory

### 3. Config File
Create a vectorless.toml file with full configuration.

## API Reference

### Engine
The main entry point for vectorless.

### IndexContext
Context for indexing documents from various sources.
"""
    ctx = IndexContext.from_text(content, name="docs", format="markdown")
    doc_id = engine.index(ctx)
    print(f"✓ Indexed: {doc_id}\n")

    # Query
    result = engine.query(doc_id, "What configuration methods are available?")
    print("Query: What configuration methods are available?")
    print(f"Score: {result.score:.2f}")
    print(f"Result: {result.content[:200]}...\n")

    # Cleanup
    engine.remove(doc_id)
    print("✓ Cleaned up")

    print("\n" + "=" * 60)
    print("Configuration Options")
    print("=" * 60)
    print()
    print("Configuration Priority (later overrides earlier):")
    print("  1. Default configuration")
    print("  2. Auto-detected config file (vectorless.toml, config.toml)")
    print("  3. Explicit config file (config_path parameter)")
    print("  4. Environment variables")
    print("  5. Constructor parameters (api_key, model, etc.)")
    print()
    print("Environment Variables:")
    print("  OPENAI_API_KEY       - LLM API key")
    print("  VECTORLESS_MODEL     - Default model name")
    print("  VECTORLESS_ENDPOINT  - API endpoint URL")
    print("  VECTORLESS_WORKSPACE - Workspace directory")
    print()
    print("Usage Examples:")
    print()
    print("# Zero configuration (recommended for beginners)")
    print('engine = Engine(workspace="./data")')
    print()
    print("# With custom model")
    print('engine = Engine(workspace="./data", model="gpt-4o-mini")')
    print()
    print("# With full config file (advanced)")
    print('engine = Engine(config_path="./vectorless.toml")')
    print()
    print("# Override config with parameters")
    print('engine = Engine(')
    print('    config_path="./vectorless.toml",')
    print('    model="gpt-4o",  # Override model from config')
    print(')')

    print("\n=== Done ===")


if __name__ == "__main__":
    main()

File: examples/rust/advanced.rs
// Copyright (c) 2026 vectorless developers
// SPDX-License-Identifier: Apache-2.0

//! Advanced usage example - Full Configuration.
//!
//! This example demonstrates how to use a full configuration file
//! for advanced use cases where you need fine-grained control.
//!
//! # Usage
//!
//! ```bash
//! # First, copy the example config and edit it
//! cp config.toml ./my_vectorless.toml
//! # Edit my_vectorless.toml to customize settings
//!
//! cargo run --example advanced
//! ```

use vectorless::{EngineBuilder, IndexContext};

#[tokio::main]
async fn main() -> vectorless::Result<()> {
    println!("=== Vectorless Advanced Example (Full Configuration) ===\n");

    // Method 1: Use explicit config file path
    // This loads all settings from the specified config file
    let client = EngineBuilder::new()
        .with_config_path("./config.toml") // or "./my_vectorless.toml"
        .build()
        .await
        .map_err(|e: vectorless::BuildError| vectorless::Error::Config(e.to_string()))?;

    println!("✓ Client created with config file\n");

    // Index a document
    let doc_id = client.index(IndexContext::from_path("./README.md")).await?;
    println!("✓ Indexed: {}\n", doc_id);

    // Query
    let result = client.query(&doc_id, "What features does Vectorless provide?").await?;
    println!("Query: What features does Vectorless provide?");
    println!("Score: {:.2}", result.score);
    if !result.content.is_empty() {
        let preview: String = result.content.chars().take(200).collect();
        println!("Result: {}...\n", preview);
    }

    // Cleanup
    client.remove(&doc_id).await?;
    println!("✓ Cleaned up");

    println!("\n=== Configuration Options ===\n");
    println!("Configuration Priority (later overrides earlier):");
    println!("  1. Default configuration");
    println!("  2. Auto-detected config file (vectorless.toml, config.toml, .vectorless.toml)");
    println!("  3. Explicit config file (with_config_path)");
    println!("  4. Environment variables (OPENAI_API_KEY, VECTORLESS_MODEL, etc.)");
    println!("  5. Builder methods (with_openai, with_model, etc.)");
    println!();
    println!("Environment Variables:");
    println!("  OPENAI_API_KEY       - LLM API key");
    println!("  VECTORLESS_MODEL     - Default model name");
    println!("  VECTORLESS_ENDPOINT  - API endpoint URL");
    println!("  VECTORLESS_WORKSPACE - Workspace directory");

    println!("\n=== Done ===");
    Ok(())
}
